Service Provider creation without entity descriptor in config file

flyinghermit opened the pull request:
- Tries to fetch entity descriptor (aka SP metadata) from remote entity ID endpoint. Requires entity ID to be configured.
- If entity descriptor is not available from remote entity ID endpoint, a default entity descriptor with entity ID, ACS URL and unspecified NameID format will be created. Requires entity ID and ACS URL to be configured. 
- New proto field ACSURL (`acs_url`) is added to accept Assertion Consumer Service URL.

see https://github.com/gravitational/teleport/issues/34661,  https://github.com/gravitational/teleport.e/issues/2692

Manually tested with service providers.

Works only with `tctl`, Web UI will be updated with another PR.

To test this PR, create a new service provider with the following spec:
```yml
kind: saml_idp_service_provider
version: v1
metadata:
  name: sp
spec:
  entity_id: https://example.com/saml/metadata
  acs_url: https://example.com/saml/acs
```

It should generate entity descriptor with the following XML format:
```xml
<EntityDescriptor xmlns="urn:oasis:names:tc:SAML:2.0:metadata" validUntil="2023-12-10T03:30:20.578Z" entityID="https://example.com/saml/metadata">
      <SPSSODescriptor xmlns="urn:oasis:names:tc:SAML:2.0:metadata" validUntil="2023-12-10T03:30:20.577825Z" protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol" AuthnRequestsSigned="false" WantAssertionsSigned="true">
        <NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified</NameIDFormat>
        <AssertionConsumerService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST" Location="https://example.com/saml/acs" index="1"></AssertionConsumerService>
        <AssertionConsumerService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Artifact" Location="https://example.com/saml/acs" index="2"></AssertionConsumerService>
      </SPSSODescriptor>
</EntityDescriptor>
```
---
Changelog: Added guided SAML entity descriptor creation when entity descriptor XML is not yet available.

flyinghermit committed:
proto def

flyinghermit committed:
fetch or generate entity descriptor
- fetch ed from remote entity ID endpoint
- if ed is not available, generate a default ed with entity ID and acs url

flyinghermit committed:
- move entity descriptor fetch and generate funcs to lib/services/local
- fetch and generate func tests
- add logger and http client to SAMLIdPServiceProviderService

flyinghermit committed:
Merge branch 'master' of github.com:gravitational/teleport into sshah/fetch-sp-metada

flyinghermit committed:
revert e ref

flyinghermit committed:
run fix-imports, fixed typos

bl-nero reviewed (approved):
The code looks good, but I was unable to properly test it (my local Teleport suddenly stopped receiving features from Sales Center and I can't force it to enable SAML).

flyinghermit commented on lib/services/local/saml_idp_service_provider.go:
```diff
@@ -183,3 +208,67 @@ func validateSAMLIdPServiceProvider(sp types.SAMLIdPServiceProvider) error {
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return trace.NotFound("not found")
```
Quick trip to the codebase I could not find one :(

flyinghermit replied:
Thanks @espadolini https://github.com/gravitational/teleport/pull/35454/commits/2a20721c4b4a535b095a2837ca60b9d05864fbdd

flyinghermit commented:
Friendly ping @r0mant @mdwn

mdwn commented on lib/services/local/saml_idp_service_provider.go:
```diff
@@ -183,3 +208,67 @@ func validateSAMLIdPServiceProvider(sp types.SAMLIdPServiceProvider) error {
+// from remote metadata endpoint (Entity ID) and sets it to sp if the xml format
+// is a valid Service Provider metadata format.
+func (s *SAMLIdPServiceProviderService) fetchAndSetEntityDescriptor(sp types.SAMLIdPServiceProvider) error {
+	var httpClient *http.Client
+	if s.httpClient != nil {
+		httpClient = s.httpClient
+	} else {
+		httpClient = &http.Client{
+			Timeout: defaults.HTTPRequestTimeout,
+			CheckRedirect: func(req *http.Request, via []*http.Request) error {
+				return http.ErrUseLastResponse
+			},
+		}
+	}
```
It feels like this bit should go in `NewSAMLIdPServiceProviderService` so that we don't have to create a new one each time given that nothing actually supplies this at present.

flyinghermit replied:
Oh yes, thanks for pointing out. I started here, then added `*http.Client` to `NewSAMLIdPServiceProviderService` but forgot to move this up there.

flyinghermit replied:
Moved to `NewSAMLIdPServiceProviderService ` in https://github.com/gravitational/teleport/pull/35454/commits/2a20721c4b4a535b095a2837ca60b9d05864fbdd

flyinghermit committed:
- return trace.ReadError for status not found
- move http client init to NewSAMLIdPServiceProviderService
- catch eof error in validateSAMLIdPServiceProvider for verbosity

r0mant commented on lib/services/local/saml_idp_service_provider.go:
```diff
@@ -40,11 +50,24 @@ const (
 
 // NewSAMLIdPServiceProviderService creates a new SAMLIdPServiceProviderService.
-func NewSAMLIdPServiceProviderService(backend backend.Backend) (*SAMLIdPServiceProviderService, error) {
+func NewSAMLIdPServiceProviderService(backend backend.Backend, httpClient *http.Client) (*SAMLIdPServiceProviderService, error) {
```
Instead of changing the function signature and having to pass `nil` everywhere, can we use functional arguments? So only callers that need to provide a custom http client will use it.

r0mant replied:
@flyinghermit Sorry if I was unclear, you would still change the function signature to accept a list of functional options but clients that don't want to provide it won't need to do so. Something like this:

```go
type SAMLIdPOption func(*SAMLIdPServiceProviderService)

func WithHTTPClient(httpClient *http.Client) SAMLIdPOption{
  return func(s *SAMLIdPServiceProviderService) {
    s.httpClient = httpClient
  }
}

func NewSAMLIdPServiceProviderService(backend backend.Backend, opts ...SAMLIdPOption) xxx {
}

// The caller that doesn't need to set custom client, nothing changes:
NewSAMLIdPServiceProviderService(backend)

// The caller that needs to set custom httpClient:
NewSAMLIdPServiceProviderService(backend, WithHTTPClinet(httpClient))
```

Does that make sense?

flyinghermit replied:
Oh got it. That would also make it easy to add/change new options in the future.

flyinghermit replied:
added in https://github.com/gravitational/teleport/pull/35454/commits/8fa2923428f124ae405a84966e886658e10815d1. Thanks for flagging, its much cleaner now!

flyinghermit committed:
Merge branch 'master' of github.com:gravitational/teleport into sshah/fetch-sp-metada

flyinghermit committed:
revert e-ref

flyinghermit committed:
add functional options to NewSAMLIdPServiceProviderService

flyinghermit committed:
remove redundant require.Error(t, err) while using require.ErrorIs

r0mant commented on lib/services/local/saml_idp_service_provider.go:
```diff
@@ -40,11 +50,22 @@ const (
+	httpClient *http.Client
+}
+
+type SAMLIdPOption func(*SAMLIdPServiceProviderService)
+
+func WithHTTPClient(httpClient *http.Client) SAMLIdPOption {
```
Nit: All public types/methods should have godocs.

r0mant commented on lib/services/local/saml_idp_service_provider.go:
```diff
@@ -40,11 +50,22 @@ const (
 // NewSAMLIdPServiceProviderService creates a new SAMLIdPServiceProviderService.
-func NewSAMLIdPServiceProviderService(backend backend.Backend) (*SAMLIdPServiceProviderService, error) {
+func NewSAMLIdPServiceProviderService(backend backend.Backend, opts ...SAMLIdPOption) (*SAMLIdPServiceProviderService, error) {
+
```
```suggestion
```

r0mant commented on lib/services/local/saml_idp_service_provider.go:
```diff
@@ -183,3 +237,60 @@ func validateSAMLIdPServiceProvider(sp types.SAMLIdPServiceProvider) error {
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return trace.Wrap(trace.ReadError(resp.StatusCode, nil))
```
Should we read response body and include it in the error and/or log it for easier troubleshooting?

r0mant reviewed (approved):
lgtm with a couple of suggestions

flyinghermit committed:
godoc for SAMLIdPOption and WithHTTPClient
